docs/enhancements/streamlined-templates.md

```diff
@@ -0,0 +1,214 @@
+# Streamlined Templates
+
+## How it works now
+
+Currently, Sonobuoy plugins are large, sprawling YAML files. This works, but it
+makes the boundary between Sonobuoy and the plugin very messy.
+
+By way of illustration, here is the current `heptio-e2e` plugin file. Every part
+of user configuration is in green, all boilerplate required by Sonobuoy is in
+red.
+
+```diff
+- ---
+- apiVersion: v1
+- kind: Pod
+- metadata:
+-   annotations:
++     sonobuoy-driver: Job
++     sonobuoy-plugin: heptio-e2e
++     sonobuoy-result-type: heptio-e2e
+-   labels:
+-     component: sonobuoy
+-     sonobuoy-run: '{{.SessionID}}'
+-     tier: analysis
+-   name: sonobuoy-heptio-e2e-job-{{.SessionID}}
+-   namespace: '{{.Namespace}}'
+- spec:
+-   containers:
++   - image: gcr.io/heptio-images/heptio-e2e:master
++     imagePullPolicy: Always
++     name: heptio-e2e
+-     volumeMounts:
+-     - mountPath: /tmp/results
+-       name: results
+-       readOnly: false
+-   - command:
+-     - sh
+-     - -c
+-     - /sonobuoy worker global -v 5 --logtostderr
+-     env:
+-     - name: NODE_NAME
+-       valueFrom:
+-         fieldRef:
+-           fieldPath: spec.nodeName
+-     - name: RESULTS_DIR
+-       value: /tmp/results
+-     - name: MASTER_URL
+-       value: '{{.MasterAddress}}'
+-     - name: RESULT_TYPE
+-       value: heptio-e2e
+-     image: gcr.io/heptio-images/sonobuoy:master
+-     imagePullPolicy: Always
+-     name: sonobuoy-worker
+-     volumeMounts:
+-     - mountPath: /tmp/results
+-       name: results
+-       readOnly: false
+-   restartPolicy: Never
+-   serviceAccountName: sonobuoy-serviceaccount
+-   tolerations:
+-   - effect: NoSchedule
+-     key: node-role.kubernetes.io/master
+-     operator: Exists
+-   - key: CriticalAddonsOnly
+-     operator: Exists
+-   volumes:
+-   - emptyDir: {}
+-     name: results
+```
+
+## Summary
+
+Removing this boiler plate will not just greatly simplify the process of making
+a plugin. It will also allow us to make changes behind the scenes to the
+Sonobuoy execution model without requiring updates to template files.
+
+All this boiler plate isn't just annoying. It paralyses changes to the Sonobuoy
+execution model. Adding
+
+## Objectives
+
+Replace the existing templates with a greatly streamlined one.
+
+### Goals
+* Conceal the execution mechanics from users
+* Allow seamless upgrading of Sonobuoy without user intervention
+
+### Non-goals
+* Do not modify the done contract betweens producer and consumer container
+
+## Proposals
```

Comment:
Couple of ideas that sparked out of this:

1) What if we did away with requiring the plugin author to write to a done file but instead told them to write to stdout. We could run their container in a dind container and wrap up the done-file semantics in that. That would be a nice way to hide those nasty details.

```yaml
spec:
- name: plugin
  image: heptio.com/sonobuoy-plugin-runner:latest
  args:
  - example.com/custom-sonobuoy-plugin:latest
  - any other
  - arguments to the custom plugin
```

2) Could we use a CRD? I don't have much experience here but what if we defined a CRD and some api aggregator that we cleaned up after a run. That might make sonobuoy a bit more flexible for us without needing any external facing templating.